openml/flows/sklearn_converter.py

```diff
@@ -0,0 +1,488 @@
+
+            if isinstance(rval, tuple):
+                parameter_value = tuple(parameter_value)
+
+            # Here (and in the elif and else branch below) are the only
+            # places where we encode a value as json to make sure that all
+            # parameter values still have the same type after
+            # deserialization
+            parameter_value = json.dumps(parameter_value)
+            parameters[k] = parameter_value
+
+        elif isinstance(rval, OpenMLFlow):
+
+            # A subcomponent, for example the base model in
+            # AdaBoostClassifier
+            sub_components[k] = rval
+            sub_components_explicit.add(k)
+            component_reference = OrderedDict()
+            component_reference['oml-python:serialized_object'] = 'component_reference'
+            component_reference['value'] = OrderedDict(key=k, step_name=None)
+            component_reference = sklearn_to_flow(component_reference)
+            parameters[k] = json.dumps(component_reference)
+
+        else:
+
+            # a regular hyperparameter
+            if not (hasattr(rval, '__len__') and len(rval) == 0):
+                rval = json.dumps(rval)
+                parameters[k] = rval
+            else:
+                parameters[k] = None
+
+        parameters_meta_info[k] = OrderedDict((('description', None),
+                                               ('data_type', None)))
+
+    # Check that a component does not occur multiple times in a flow as this
+    # is not supported by OpenML
+    to_visit_stack = []
+    to_visit_stack.extend(sub_components.values())
+    known_sub_components = set()
+    while len(to_visit_stack) > 0:
+        visitee = to_visit_stack.pop()
+        if visitee.name in known_sub_components:
+            raise ValueError('Found a second occurence of component %s when '
+                             'trying to serialize %s.' % (visitee.name, model))
+        else:
+            known_sub_components.add(visitee.name)
+            to_visit_stack.extend(visitee.components.values())
+
+    # Create a flow name, which contains all components in brackets, for
+    # example RandomizedSearchCV(Pipeline(StandardScaler,AdaBoostClassifier(DecisionTreeClassifier)),StandardScaler,AdaBoostClassifier(DecisionTreeClassifier))
+    class_name = model.__module__ + "." + model.__class__.__name__
+
+    # will be part of the name (in brackets)
+    sub_components_names = ""
+    for key in sub_components:
+        if key in sub_components_explicit:
+            sub_components_names += "," + key + "=" + sub_components[key].name
+        else:
+            sub_components_names += "," + sub_components[key].name
+
+    if sub_components_names:
+        # slice operation on string in order to get rid of leading comma
+        name = '%s(%s)' % (class_name, sub_components_names[1:])
+    else:
+        name = class_name
+
+    # Get the external versions of all sub-components
```

Comment:
Extracted a function here, as well as for checking that a component is not used multiple times in a flow.